pytest/tests/mirror/test.py

```diff
@@ -0,0 +1,316 @@
+import shutil
+import signal
+import subprocess
+
+sys.path.append(str(pathlib.Path(__file__).resolve().parents[2] / 'lib'))
+
+from cluster import init_cluster, spin_up_node, load_config
+from configured_logger import logger
+from mocknet import create_genesis_file
+import transaction
+import utils
+import key
+
+# This sets up an environment to test the tools/mirror process. It starts a localnet with a few validators
+# and waits for some blocks to be produced. Then we fork the state and start a new chain from that, and
+# start the mirror process that should mirror transactions from the source chain to the target chain.
+# Halfway through we restart it to make sure that it still works properly when restarted
+
+TIMEOUT = 240
+NUM_VALIDATORS = 4
+SHARDNET_VALIDATORS = ['foo0', 'foo1', 'foo2']
+
+
+def mkdir_clean(dirname):
+    try:
+        os.mkdir(dirname)
+    except FileExistsError:
+        shutil.rmtree(dirname)
+        os.mkdir(dirname)
+
+
+def dot_near():
+    return pathlib.Path.home() / '.near'
+
+
+def init_shardnet_dir(neard, home, ordinal, validator_account=None):
+    mkdir_clean(home)
+
+    process = subprocess.Popen([neard, '--home', home, 'init'],
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.PIPE)
+    (_, err) = process.communicate()
+    if process.returncode != 0:
+        sys.exit(
+            f'"neard init" command returned {process.returncode}: stderr: {err}'
+        )
+    shutil.copy(dot_near() / 'test0/config.json', home / 'config.json')
+    shutil.copy(dot_near() / 'test0/forked/genesis.json', home / 'genesis.json')
+    shutil.copy(dot_near() / 'test0/forked/records.json', home / 'records.json')
+
+    with open(home / 'config.json', 'r') as f:
+        config = json.load(f)
+        config['genesis_records_file'] = 'records.json'
+        config['network']['addr'] = f'0.0.0.0:{24567 + 10 + ordinal}'
+        config['rpc']['addr'] = f'0.0.0.0:{3030 + 10 + ordinal}'
+    with open(home / 'config.json', 'w') as f:
+        json.dump(config, f)
+
+    if validator_account is None:
+        os.remove(home / 'validator_key.json')
+    else:
+        # this key and the suffix -load-test.near are hardcoded in create_genesis_file()
+        with open(home / 'validator_key.json', 'w') as f:
+            f.write('{\n')
+            f.write(
+                f'"account_id": "{validator_account + "-load-test.near"}",\n')
+            f.write(
+                '"public_key": "ed25519:76NVkDErhbP1LGrSAf5Db6BsFJ6LBw6YVA4BsfTBohmN",\n'
+            )
+            f.write(
+                '"secret_key": "ed25519:3cCk8KUWBySGCxBcn1syMoY5u73wx5eaPLRbQcMi23LwBA3aLsqEbA33Ww1bsJaFrchmDciGe9otdn45SrDSkow2"\n'
+            )
+            f.write('}\n')
+
+
+def init_shardnet_dirs(neard):
+    ordinal = NUM_VALIDATORS + 1
+    dirs = []
+
+    for account_id in SHARDNET_VALIDATORS:
+        home = dot_near() / f'shardnet_{account_id}'
+        dirs.append(str(home))
```

Comment:
It's because of the `atexit.cleanup()` here: https://github.com/near/nearcore/blob/5532a1ff8544fd4708618d60a950a30c7b1f09f5/pytest/lib/cluster.py#L511